File: Tree/236_LowestCommonAncestorOfBinaryTree.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    """
    Recursive method: DFS.
    If the current (sub)tree contains both p and q, then the function result is their LCA.
    If only one of them is in that subtree, then the result is that one of them.
    If neither are in that subtree, the result is null/None/nil.

    More version can be found here:
    https://discuss.leetcode.com/topic/18561/4-lines-c-java-python-ruby
    """
    def lowestCommonAncestor(self, root, p, q):
        if not root or root == p or root == q:
            return root
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)
        # if p and q are on both sides
        if left and right:
            return root
        else:
            return left or right


class Solution_2(object):
    """
    Iterative method: BFS(DFS is ok too).  According to:
    https://leetcode.com/discuss/64764/java-python-iterative-solution
    """
    def lowestCommonAncestor(self, root, p, q):
        node_stack = [root]
        parent_record = {root: None}

        # Build the relationship from child to parent
        while p not in parent_record or q not in parent_record:
            node = node_stack.pop()
            if node.left:
                node_stack.append(node.left)
                parent_record[node.left] = node
            if node.right:
                node_stack.append(node.right)
                parent_record[node.right] = node

        # Trace brack from one node, record the path.
        # Then trace from the other.
        ancestors = set()
        while p:
            ancestors.add(p)
            p = parent_record[p]

        while q not in ancestors:
            q = parent_record[q]
        return q

File: Tree/236_LowestCommonAncestorOfBinaryTree.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-09-16 12:35:32
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    /*
    Recursive method: DFS.
    If the current (sub)tree contains both p and q, then the function result is their LCA.
    If only one of them is in that subtree, then the result is that one of them.
    If neither are in that subtree, the result is null/None/nil.

    More version can be found here:
    https://discuss.leetcode.com/topic/18561/4-lines-c-java-python-ruby
    */
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root==NULL || root==p || root==q){
            return root;
        }
        TreeNode* l = lowestCommonAncestor(root->left, p, q);
        TreeNode* r = lowestCommonAncestor(root->right, p, q);
        if(l && r){
            return root;
        }
        else{
            return l ? l : r;
        }
    }
};

class Solution_2 {
public:
    /*
    Iteratice method.
    */
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        map<TreeNode*, TreeNode*> parent;
        parent.insert({root, NULL});
        queue<TreeNode*> que;
        que.push(root);
        // Build the relationship from child to parent
        while(parent.find(p) == parent.end() || parent.find(q) == parent.end()){
            TreeNode *cur = que.front();
            que.pop();
            if(cur->left){
                que.push(cur->left);
                parent.insert({cur->left, cur});
            }
            if(cur->right){
                que.push(cur->right);
                parent.insert({cur->right, cur});
            }
        }
        // Trace brack from one node, record the path. Then trace from the other.
        set<TreeNode*> ancestor;
        while(p){
            ancestor.insert(p);
            p = parent[p];
        }
        while(ancestor.find(q) == ancestor.end()){
            q = parent[q];
        }
        return q;
    }
};

